docs/zkapps/zkapp-development-frameworks.mdx

```diff
@@ -15,68 +15,214 @@ keywords:
+
+Protokit can operate in two modes: Based Sequencing and Hybrid Sequencing.
+
+In **Based Sequencing mode**, Protokit sends all transactions to the Mina L1, providing the decentralization guarantees of the L1, but limited to the L1’s throughput and latency. Based Sequencing mode is best for zkApps where censorship resistance and liveness are more important than the additional performance benefits.
+
+In **Hybrid Sequencing mode**, Protokit runs an off-chain sequencer with off-chain data availability, providing higher throughput and lower latency than the Mina L1 but at the expense of decentralization. Hybrid Sequencing mode is best for zkApps where performance is critical.
+
+**Important:** Users of a zkApp in Hybrid Sequencing mode can always **submit transactions directly to the L1**, for example, to withdraw funds (in case the zkApp sequencer is unavailable) but it does rely on the Protokit’s data availability (DA). Solutions are in the works for DA guarantees even in Hybrid Sequencing mode so that developers can get higher throughput and latency while retaining censorship resistance and liveness guarantees.
+
+Start developing zkApps with Protokit here:
 
-o1js is also the **zkDSL** used for:
 
-- Writing general-purpose zk circuits.
-- Constructing new primitives and data structures.
 
-There are some key considerations when choosing to build a zkApp with o1js on Mina L1:
+* [Developer documentation](https://protokit.dev/docs/what-is-protokit)
+* [Protokit repository](https://github.com/proto-kit)
+* [Starter Kit](https://github.com/proto-kit/starter-kit)
 
-- zkApps are subject to protocol throughput limitations.
-- At present, zkApps that require support for multiple concurrent users require specific architecture to avoid race conditions:
-  - Where more than the eight on-chain field elements are required to manage state, and access to that state is not shared between users, the experimental [Offchain Storage API](/zkapps/writing-a-zkapp/feature-overview/offchain-storage) offers a solution.
-  - Where concurrent access to _shared global state_ is required, the required architecture is available **out of the box** when using the Protokit framework to build your zkApp as an zkApp-chain (L2). There is currently no easy-to-use equivalent for shared state in o1js L1 contracts.
 
-Start here:
+## o1js[​](https://docs.minaprotocol.com/zkapps/zkapp-development-frameworks#o1js)
 
-- [Developer documentation](/zkapps/o1js)
-- [o1js repository](https://github.com/o1-labs/o1js)
-- [Discord](https://discord.gg/minaprotocol)
+[o1js](https://docs.minaprotocol.com/zkapps/o1js) is the best way to author **highly-optimized ZK circuits and primitives**, but it does not provide an easy-to-use solution for managing a shared global state.
```

Comment:
`OffchainState` provides an out-of-the box solution to this. It's worth mentioning here.

Reply:
I don't think we should highlight offchain state as a serious way to handle app state until it's out of the experimental namespace.  I don't think we should make casual references to things in the docs until we're ready to stand behind them 100%.

Reply:
Can we consider this closed? Jan has address Florian's comment and we're not going to include mention of OffchainState.

Reply:
Its not true that OffchainState is a solution to race conditions on "shared global state" as suggested above. It only solves this for user-specific state. The thing is that you can't concurrently compute over shared global state without any form of on-chain execution. That is why we built protokit, to add that possibility to the features already existing in o1js.
If you wanted to do the same with vanilla-o1js features and then polish the DX, solve other limitations in the process and then make the system efficient, you'd pretty much end up with protokit in based sequencing mode (which is the first mode we will ship)